Implementation of MK docs. (#33)
FAIRmat-NFDI/nyaml: docs/learn/explanations.md

Reviewer: hyperlink to https://yaml.org/ when writing YAML upon first occurance

Reviewer: Was this addressed?

Author: Addressed!

# NeXus Definition Language (NXDL), XML, and YAML

> The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/) supporting FAIR (Findable, Accessible, Interoperable and Reuseable) data principles in material science. The format serves as a standardized framework for both data exchange and storage.

Reviewer 2: ```suggestion
The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research (see the most recent NeXus paper by [Könnecke et al.](https://doi.org/10.1107/S1600576714027575)). The data format is also used by the material science community under the project [NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/) supporting FAIR (Findable, Accessible, Interoperable and Reuseable) data principles in material science. The format serves as a standardized framework for both data exchange and storage.
```

Author: Addressed!

> The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/) supporting FAIR (Findable, Accessible, Interoperable and Reuseable) data principles in material science. The format serves as a standardized framework for both data exchange and storage.

Reviewer: Supporting the rewriting, we must not point though to outdated links.
Therefore, use [FAIRmat-NFDI ](https://github.com/FAIRmat-NFDI/nexus_definitions) instead of the fairmat-experimental -ish

Author: I'm not sure why the `NeXus-FAIRmat` is outdated. The platform where the appdef and base classes are being represented from `nexus-definition` repo.

Reviewer: Irrelevant why they are outdated, they are, please update the link to as you said the https://fairmat-nfdi.github.io/nexus_definitions/, comment and resolve

> The NeXus data format, described by the NeXus Definition Language (NXDL), represents a concerted effort aimed at facilitating data exchange within scientific communities, particularly among those engaged in neutron, X-ray, and muon research [Könnecke, M. et al.](https://doi.org/10.1107/S1600576714027575). The data format is also being used by the material science community under the project [NeXus-FAIRmat proposal](https://fairmat-experimental.github.io/nexus-fairmat-proposal/) supporting FAIR (Findable, Accessible, Interoperable and Reuseable) data principles in material science. The format serves as a standardized framework for both data exchange and storage.

Reviewer: "Interoperable, and Reusable"

Reviewer: Drop "data" in front of "principles"

Reviewer: Add Wilkinson paper when mentioning FAIR
https://www.nature.com/articles/sdata201618

Author: Addressed!

> At the NeXus data format's core, the NeXus Definition Language (NXDL) functions as the cornerstone through which scientists delineate the nomenclature and organizational structure of information within NeXus data files, tailored to specific scientific techniques. NXDL is used to define general data structures (base classes), which define the set of terms that might be used in an instance of that class. These base classes are the building blocks for defining, under the schema called application definition, measurement-specific or even instrument-specific or software-specific data storage objects. In this process, members and definitions of individual base classes can be used as is or customized. In essence, the process of schema development, whether for a base class or an application definition, entails crafting an NXDL schema definition file with the extension 'nxdl.xml', utilizing the Extensible Markup Language, [XML](https://www.w3.org/TR/REC-xml/REC-xml-20081126.xml).

Reviewer: "The NeXus Definition Language (NXDL) at the core of the NeXus data format functions as ..."

Reviewer: "cornerstone through which" > "whereby"

Reviewer: pieces of information

Reviewer: personally would remove ", tailored to ..." cuz NXDL is always used irrespective if class was tailored or not

Reviewer: remove comma before [XML]

Author: addressed!

> To expedite the schema development process, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within the schema. The YAML format reduces repetition of NXDL keywords and offers a more intuitive structure aligned with object-oriented programming concepts, such as class inheritance. These benefits are attained without compromising the integrity of the original NeXus schema, which is traditionally expressed in XML format.

Reviewer: "we have introduced the use of" > "recent the use of ... was introduced"

"The YAML format reduces repetition of NXDL keywords and offers a" > "Thus, using YAML reduces repetition of NXDL keywords and offers a"

Author: Addressed!

> To expedite the schema development process, recently, we have introduced the use of Yet Another Markup Language ([YAML](https://yaml.org/)), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within the schema. Thus, the use of YAML format reduces repetition of NXDL keywords and offers a more intuitive structure aligned with object-oriented programming concepts, such as class inheritance. These benefits are attained without compromising the integrity of the original NeXus schema, which is traditionally expressed in XML format.

Reviewer 2: It isn't called Yet Another Markup Language anymore, maybe just use the acronym
```suggestion
To expedite the schema development process, recently, we have introduced the use of [YAML](https://yaml.org/), which provides a syntax or style specifically tailored for defining scientific domain-driven schemas with NXDL. One significant advantage of YAML over XML is its indentation-driven approach, which eliminates the need for starting and ending tags for each entity within the schema. Thus, the use of YAML format reduces repetition of NXDL keywords and offers a more intuitive structure aligned with object-oriented programming concepts, such as class inheritance. These benefits are attained without compromising the integrity of the original NeXus schema, which is traditionally expressed in XML format.
```

Reviewer: Trivial thing, fixed? if so, mention and resolve

> While `YAML` offers a more readable and concise way to author NeXus definitions, a transcoding mechanism is required because the official format remains XML. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml) Python package serves as a converter tool designed specifically for this purpose. It enables the conversion of NXDL-compliant definitions from YAML to XML, making it easier for NeXus schema developers to create and maintain definitions. Furthermore, the tool offers the flexibility to extend existing NeXus schemas in XML by facilitating conversion back and forth between the two formats. It is important to note that here we do not introduce NeXus data objects, terms, or types, which are fundamental for writing base class schemas or application definition schemas. For individuals new to NeXus, we refer to the official NeXus site at NeXus ([NeXus official website](https://www.nexusformat.org/)).

Reviewer: "designed specifically for" > "for specifically"

Author: Addressed!

> While `YAML` offers a more readable and concise way to author NeXus definitions, a transcoding mechanism is required because the official format remains XML. The [nyaml](https://github.com/FAIRmat-NFDI/nyaml) Python package serves as a converter tool designed for specifically this purpose. It enables the conversion of NXDL-compliant definitions from YAML to XML, making it easier for NeXus schema developers to create and maintain definitions. Furthermore, the tool offers the flexibility to extend existing NeXus schemas in XML by facilitating conversion back and forth between the two formats. It is important to note that here we do not introduce NeXus data objects, terms, or types, which are fundamental for writing base class schemas or application definition schemas. For individuals new to NeXus, we refer to the official NeXus site at NeXus ([NeXus official website](https://www.nexusformat.org/)).

Reviewer 2: Decide if you want to write plain "YAML" or "`YAML`"

Reviewer: This is a point about consistent writing, trivial fix:
1.) Make a choice e.g. `YAML`
2.) Search for occurrences and replace by above-choice
3.) Leave a comment like fixed and done

Author: Kept `YAML` and fixed the inconsistency over the md files.
